File: Java/ExtensibleCommands/src/main/java/org/extensiblecommands/ExtensibleCommandsException.java
package org.extensiblecommands;

/**
 * Basic type for all operational exceptions thrown from within command execution path 
 */
public class ExtensibleCommandsException extends Exception {
    private final int id;
    private final String text;
    private Exception exception;
    private final String timestamp;

    /**
     * Constructor
     * @param id            Error ID
     * @param text          Error description
     */
    public ExtensibleCommandsException(int id, String text) {
        this.id = id;
        this.text = text;
        this.timestamp = Logger.getTimeStamp();
    }

    /**
     * Constructor
     * @param id            Error ID
     * @param text          Error description
     * @param exception     Underlying exception
     */
    public ExtensibleCommandsException(int id, String text, Exception exception) {
        this(id, text);
        this.exception = exception;
    }

    /**
     * @return              Error ID
     */
    public int getId() {
        return id;
    }

    /**
     * @return              Error description
     */
    public String getText() {
        return text;
    }

    /**
     * @return              Underlying exception
     */
    public Exception getException() {
        return exception;
    }

    /**
     * @return              Timestamp of when the failure occurred
     */
    public String getTimestamp() {
        return timestamp;
    }
}

File: CSharp/ExtensibleCommands/ExtensibleCommands/ExtensibleCommandsException.cs
using System;

namespace ExtensibleCommands
{
    /// <summary> 
    /// Basic type for all operational exceptions thrown from within command execution path 
    /// </summary>
    public class ExtensibleCommandsException : Exception
    {
        /// <summary> Error ID </summary>
        public int ID { get; }

        /// <summary> Error description </summary>
        public string Text { get; }

        /// <summary> Underlying exception </summary>
        public Exception Exception { get; }

        /// <summary> Timestamp of when the failure occurred </summary>
        public string Timestamp { get; }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="id">Error ID</param>
        /// <param name="text">Error description</param>
        public ExtensibleCommandsException(int id, string text)
            : base(text)
        {
            ID              = id;
            Text            = text;
            Timestamp       = Logger.TimeStamp;
        }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="id">Error ID</param>
        /// <param name="text">Error description</param>
        /// <param name="e">Underlying exception</param>
        public ExtensibleCommandsException(int id, string text, Exception e)
            : this(id, text)
        {
            Exception = e;
        }
    }
}
